source/dshow-settings.cpp

```diff
@@ -0,0 +1,303 @@
+                    if (SUCCEEDED(hr)) {
+                        settings.emplace("Exposure", value);
+                        settings.emplace("Exposure_flag", (flag > CameraControl_Flags_Auto ? "manual" : "auto"));
+                    }
+
+                    hr = camControlSettings->Get(CameraControl_Focus, &value, &flag);
+                    if (SUCCEEDED(hr)) {
+                        settings.emplace("Focus", value);
+                        settings.emplace("Focus_flag", (flag > CameraControl_Flags_Auto ? "manual" : "auto"));
+                    }
+
+                    hr = camControlSettings->Get(CameraControl_Iris, &value, &flag);
+                    if (SUCCEEDED(hr)) {
+                        settings.emplace("Iris", value);
+                        settings.emplace("Iris_flag", (flag > CameraControl_Flags_Auto ? "manual" : "auto"));
+                    }
+
+                    hr = camControlSettings->Get(CameraControl_Pan, &value, &flag);
+                    if (SUCCEEDED(hr)) {
+                        settings.emplace("Pan", value);
+                        settings.emplace("Pan_flag", (flag > CameraControl_Flags_Auto ? "manual" : "auto"));
+                    }
+
+                    hr = camControlSettings->Get(CameraControl_Roll, &value, &flag);
+                    if (SUCCEEDED(hr)) {
+                        settings.emplace("Roll", value);
+                        settings.emplace("Roll_flag", (flag > CameraControl_Flags_Auto ? "manual" : "auto"));
+                    }
+
+                    hr = camControlSettings->Get(CameraControl_Tilt, &value, &flag);
+                    if (SUCCEEDED(hr)) {
+                        settings.emplace("Tilt", value);
+                        settings.emplace("Tilt_flag", (flag > CameraControl_Flags_Auto ? "manual" : "auto"));
+                    }
+
+                    hr = camControlSettings->Get(CameraControl_Zoom, &value, &flag);
+                    if (SUCCEEDED(hr)) {
+                        settings.emplace("Zoom", value);
+                        settings.emplace("Zoom_flag", (flag > CameraControl_Flags_Auto ? "manual" : "auto"));
+                    }
+                }
+            }
+        }
+
+        camSettings.push_back(settings);
+    }
+
+    static recursive_mutex settingsMutex;
+
+    void LoadSettingsFromFile(const std::string& filePath)
+    {
+        lock_guard<recursive_mutex> lock(settingsMutex);
+
+        ComPtr<ICreateDevEnum> deviceEnum;
+        HRESULT hr;
+
+        hr = CoCreateInstance(CLSID_SystemDeviceEnum, nullptr,
+                              CLSCTX_INPROC_SERVER, IID_ICreateDevEnum, (void **)&deviceEnum);
+
+        if (SUCCEEDED(hr)) {
+            ComPtr<IEnumMoniker> enumMoniker;
+            hr = deviceEnum->CreateClassEnumerator(CLSID_VideoInputDeviceCategory, &enumMoniker, 0);
+
+            if (SUCCEEDED(hr)) {
+                std::ifstream file(filePath);
```

Comment:
How is this file being created? My understanding when reading the API is that ifstream will only open an existing file.

Reply:
The is created in the SaveSettingsToFile call (line 274). The overload of the ifstream constructor that receives a string with a file name associates the stream with that file. Basically it calls open internally. With that said I see that I forgot to check if the file actually opened with is_open(). I'll add that check

Reply:
kk sounds good, I think we should handle in some ways if we try to load settings when the file has not been created -> we should create and set default settings.

Reply:
Added the is_open check. On the default values, I don't think it is safe to assume a default value for every possible device out there. Some value we choose could break something for a device. The first time you load the webcam or any other type of capture device the settings values shown there are the default that come from the device. So I think it is best to have the device give it's default values to us instead of assuming values ourselves

Reply:
Good for adding check.

Yes, agree with not setting default values our self but save whatever config is currently used by the device. Where is this logic happening? I see we are not trying to load anything now when the file doesn't exist, I think we should then get the current settings and save them on the file.

Reply:
Makes sense, I'll add that

Reply:
Added a call to SaveSettingsToFile if when trying to load the settings it is not able to open the file

Reply:
Sounds good, perfect.